runtime/src/main/java/org/corfudb/util/serializer/ProtobufSerializer.java

```diff
@@ -0,0 +1,143 @@
+package org.corfudb.util.serializer;
+
+import com.google.protobuf.Any;
+import com.google.protobuf.Message;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.ByteBufOutputStream;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.CorfuStoreMetadata.Record;
+import org.corfudb.runtime.collections.CorfuRecord;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * The Protobuf serializer is the main component that allows CorfuStore to use Protobufs to
+ * convert a language specific (Java here) object into an identifiable byte buffer that
+ * can be then converted back to a language specific object.
+ *
+ * To achieve this, this serializer requires a map of all seen class types or a classMap
+ * and Google Protobuf 3's Any.
+ * Any type carries with it a typeUrl which help identify the class uniquely.
+ * This typeUrl is then used to index the classMap to retrieve the actual Protobuf message
+ * while deserializing.
+ */
+@Slf4j
+public class ProtobufSerializer implements ISerializer {
+
+    private final byte type;
+    private final Map<String, Class<? extends Message>> classMap;
+
+    public ProtobufSerializer(byte type, Map<String, Class<? extends Message>> classMap) {
+        this.type = type;
+        this.classMap = classMap;
+    }
+
+    private enum MessageType {
+        KEY(1),
+        VALUE(2);
+
+        private static final Map<Integer, MessageType> valToTypeMap = new HashMap<>();
+
+        static {
+            for (MessageType type : MessageType.values()) {
+                valToTypeMap.put(type.val, type);
+            }
+        }
+
+        private final int val;
+
+        MessageType(int val) {
+            this.val = val;
+        }
+
+        public static MessageType valueOf(int val) {
+            return valToTypeMap.get(val);
+        }
+    }
+
+    @Override
+    public byte getType() {
+        return type;
+    }
+
+    /**
+     * Deserialize an object from a given byte buffer.
+     *
+     * @param b The bytebuf to deserialize.
+     * @return The deserialized object.
+     */
+    @Override
+    public Object deserialize(ByteBuf b, CorfuRuntime rt) {
+
+        try (ByteBufInputStream bbis = new ByteBufInputStream(b)) {
+            MessageType type = MessageType.valueOf(bbis.readInt());
```

Comment:
type could be null, right?

Reply:
should not be if data was written by one of Corfu's layers.